### ISSUE TITLE
Building Tensorflow 1.13 for python 3.8 fails on nullptr conversion.

### ISSUE THREAD

**Issue by odinsbane** [tensorflow/tensorflow#34197]
**System information**
- OS Platform and Distribution: Centos 7
- TensorFlow installed from : source
- TensorFlow version: 1.13.2
- Python version: 3.8.0
- Bazel version : 0.21.0
- GCC/Compiler version : gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39) 
- CUDA/cuDNN version: 9.1/7.0
- GPU model and memory: 1080ti

**Describe the problem**

When attempting to build the tensorflow branch r1.13 targeting python 3.8 the build fails with the following error:

ndarray_tensor_bridge.cc:108:1: error: cannot convert 'std::nullptr_t' to 'Py_ssize_t {aka long int}' in initialization
 };

This error occurs in three files that I found: bfloat16.cc, ndarray_tensor_bridge.cc, pywrap_tfe_src.cc

This appears to be the same problem fixed in issue #33543 

**Provide the exact sequence of commands / steps that you executed before running into the problem**

I am willing to provide more details, but this appears to be the exact same issue. I am including the git diff from the version that compiled and works for me.

```patch
diff --git a/tensorflow/python/eager/pywrap_tfe_src.cc b/tensorflow/python/eager/pywrap_tfe_src.cc
index 9ce500b..726c4c5 100644
--- a/tensorflow/python/eager/pywrap_tfe_src.cc
+++ b/tensorflow/python/eager/pywrap_tfe_src.cc
@@ -1216,7 +1216,7 @@ static PyTypeObject TFE_Py_Tape_Type = {
     sizeof(TFE_Py_Tape),                          /* tp_basicsize */
     0,                                            /* tp_itemsize */
     &TFE_Py_Tape_Delete,                          /* tp_dealloc */
-    nullptr,                                      /* tp_print */
+    NULL,                                         /* tp_print */
     nullptr,                                      /* tp_getattr */
     nullptr,                                      /* tp_setattr */
     nullptr,                                      /* tp_reserved */
diff --git a/tensorflow/python/lib/core/bfloat16.cc b/tensorflow/python/lib/core/bfloat16.cc
index fde3a83..e0da0f4 100644
--- a/tensorflow/python/lib/core/bfloat16.cc
+++ b/tensorflow/python/lib/core/bfloat16.cc
@@ -317,7 +317,7 @@ PyTypeObject PyBfloat16_Type = {
     sizeof(PyBfloat16),                        // tp_basicsize
     0,                                         // tp_itemsize
     nullptr,                                   // tp_dealloc
-    nullptr,                                   // tp_print
+    NULL,                                      // tp_print
     nullptr,                                   // tp_getattr
     nullptr,                                   // tp_setattr
     nullptr,                                   // tp_compare / tp_reserved
diff --git a/tensorflow/python/lib/core/ndarray_tensor_bridge.cc b/tensorflow/python/lib/core/ndarray_tensor_bridge.cc
index 0d58385..43ab92c 100644
--- a/tensorflow/python/lib/core/ndarray_tensor_bridge.cc
+++ b/tensorflow/python/lib/core/ndarray_tensor_bridge.cc
@@ -86,7 +86,7 @@ PyTypeObject TensorReleaserType = {
     0,                                /* tp_itemsize */
     /* methods */
     TensorReleaser_dealloc,      /* tp_dealloc */
-    nullptr,                     /* tp_print */
+    NULL,                        /* tp_print */
     nullptr,                     /* tp_getattr */
     nullptr,                     /* tp_setattr */
     nullptr,                     /* tp_compare */
```

**Comment by tensorfoo** [comment]
@odinsbane thanks for those diffs, i'm using 1.14 and 3.8 as well and was able to complete a successful build using your changes.

**Comment by oscarriddle** [comment]
CentOS7, TF1.14, CUDA10.1, PYTHON3.8, cuDNN7.5, TRT6
After applying this diff patch, successfully built.

**Comment by gunan** [comment]
As the old branches are closed, we will not backport the compilation fix to these branches.
Only security issues will be patched and released for old branches.

**Comment by odinsbane** [comment]
The documentation here, https://www.tensorflow.org/guide/upgrade says to
upgrade to 1.14 for 2.0 migration. Is that the recommended route to
upgrade, or is 1.15 preferred? I am not ready to port to 2.X yet, but I
would like to be ready to.

**Comment by gunan** [comment]
1.15 should be better, as there are bugs we have fixed after 1.14.

**Comment by tigerhawkvok** [comment]
For a future person, this patch sucessfully compiled on Windows for Python 3.9.5

```diff
diff --git a/tensorflow/python/eager/pywrap_tensor.cc b/tensorflow/python/eager/pywrap_tensor.cc
index b81eddac07..09a618a4d7 100644
--- a/tensorflow/python/eager/pywrap_tensor.cc
+++ b/tensorflow/python/eager/pywrap_tensor.cc
@@ -839,7 +839,7 @@ static PyTypeObject _EagerTensorType = {
     sizeof(EagerTensor),                /* tp_basicsize */
     0,                                  /* tp_itemsize */
     (destructor)EagerTensor_dealloc,    /* tp_dealloc */
-    nullptr,                            /* tp_print */
+    NULL,                            /* tp_print */
     nullptr,                            /* tp_getattr */
     nullptr,                            /* tp_setattr */
     nullptr,                            /* tp_compare */
diff --git a/tensorflow/python/eager/pywrap_tfe_src.cc b/tensorflow/python/eager/pywrap_tfe_src.cc
index ba9d8a36b0..159b91ff1b 100644
--- a/tensorflow/python/eager/pywrap_tfe_src.cc
+++ b/tensorflow/python/eager/pywrap_tfe_src.cc
@@ -1319,7 +1319,7 @@ static PyTypeObject TFE_Py_Tape_Type = {
     sizeof(TFE_Py_Tape),                          /* tp_basicsize */
     0,                                            /* tp_itemsize */
     &TFE_Py_Tape_Delete,                          /* tp_dealloc */
-    nullptr,                                      /* tp_print */
+    NULL,                                      /* tp_print */
     nullptr,                                      /* tp_getattr */
     nullptr,                                      /* tp_setattr */
     nullptr,                                      /* tp_reserved */
@@ -1353,7 +1353,7 @@ static PyTypeObject TFE_Py_ForwardAccumulator_Type = {
     sizeof(TFE_Py_ForwardAccumulator),                      /* tp_basicsize */
     0,                                                      /* tp_itemsize */
     &TFE_Py_ForwardAccumulatorDelete,                       /* tp_dealloc */
-    nullptr,                                                /* tp_print */
+    NULL,                                                /* tp_print */
     nullptr,                                                /* tp_getattr */
     nullptr,                                                /* tp_setattr */
     nullptr,                                                /* tp_reserved */
diff --git a/tensorflow/python/lib/core/bfloat16.cc b/tensorflow/python/lib/core/bfloat16.cc
index fde3a83770..488e60089a 100644
--- a/tensorflow/python/lib/core/bfloat16.cc
+++ b/tensorflow/python/lib/core/bfloat16.cc
@@ -200,7 +200,7 @@ PyNumberMethods PyBfloat16_AsNumber = {
 #if PY_MAJOR_VERSION < 3
     PyBfloat16_Int,  // nb_long
 #else
-    nullptr,  // reserved
+    NULL,  // reserved
 #endif
     PyBfloat16_Float,  // nb_float
 #if PY_MAJOR_VERSION < 3
@@ -317,10 +317,10 @@ PyTypeObject PyBfloat16_Type = {
     sizeof(PyBfloat16),                        // tp_basicsize
     0,                                         // tp_itemsize
     nullptr,                                   // tp_dealloc
-    nullptr,                                   // tp_print
+    NULL,                                   // tp_print
     nullptr,                                   // tp_getattr
     nullptr,                                   // tp_setattr
-    nullptr,                                   // tp_compare / tp_reserved
+    NULL,                                   // tp_compare / tp_reserved
     PyBfloat16_Repr,                           // tp_repr
     &PyBfloat16_AsNumber,                      // tp_as_number
     nullptr,                                   // tp_as_sequence
@@ -489,8 +489,9 @@ bool RegisterBfloat16Cast(int numpy_type, bool cast_is_safe) {
   return true;
 }
 
+// https://github.com/tensorflow/tensorflow/issues/41086#issuecomment-656833081
 template <typename InType, typename OutType, typename Functor>
-void BinaryUFunc(char** args, npy_intp* dimensions, npy_intp* steps,
+void BinaryUFunc(char** args, npy_intp const* dimensions, npy_intp const* steps,
                  void* data) {
   const char* i0 = args[0];
   const char* i1 = args[1];
@@ -505,11 +506,18 @@ void BinaryUFunc(char** args, npy_intp* dimensions, npy_intp* steps,
   }
 }
 
+// https://github.com/NVIDIA/tensorflow/commit/746a984f67109985fcffbfdf969f8706cf5925af#diff-10fde2c960bdd6b523b0d0986b5d7a9d69610012d0ac400c1cfcd3beee51e5ff
+// Numpy changed const-ness of PyUFuncGenericFunction, provide overload.
 template <typename Functor>
 void CompareUFunc(char** args, npy_intp* dimensions, npy_intp* steps,
                   void* data) {
   BinaryUFunc<bfloat16, npy_bool, Functor>(args, dimensions, steps, data);
 }
+template <typename Functor>
+void CompareUFunc(char** args, const npy_intp* dimensions,
+                  const npy_intp* steps, void* data) {
+  BinaryUFunc<bfloat16, npy_bool, Functor>(args, dimensions, steps, data);
+}
 
 struct Bfloat16EqFunctor {
   npy_bool operator()(bfloat16 a, bfloat16 b) { return a == b; }
diff --git a/tensorflow/python/lib/core/ndarray_tensor_bridge.cc b/tensorflow/python/lib/core/ndarray_tensor_bridge.cc
index 16f6934186..ae7720007e 100644
--- a/tensorflow/python/lib/core/ndarray_tensor_bridge.cc
+++ b/tensorflow/python/lib/core/ndarray_tensor_bridge.cc
@@ -86,7 +86,7 @@ PyTypeObject TensorReleaserType = {
     0,                                /* tp_itemsize */
     /* methods */
     TensorReleaser_dealloc,      /* tp_dealloc */
-    nullptr,                     /* tp_print */
+    NULL,                     /* tp_print */
     nullptr,                     /* tp_getattr */
     nullptr,                     /* tp_setattr */
     nullptr,                     /* tp_compare */
```

using `git apply --reject --whitespace=fix --ignore-space-change --ignore-whitespace patch.diff`

If you're on Windows, I wrote a build script to do it here:

[tensorflow-builder-win.zip](https://github.com/tensorflow/tensorflow/files/11763613/tensorflow-builder-win.zip)
